tests/api2/test_disk_format.py

```diff
@@ -1,118 +1,166 @@
-    data_size = disk_size - swap_size
-    call('disk.format', disk, data_size, 2)
+    # Should be a modulo of grain_size
+    assert (partitions[0]['end_sector'] - partitions[0]['start_sector']) % grain_size == 0
 
-    partitions = call('disk.list_partitions', disk)
-    assert len(partitions) == 2
-    # As much swap as we could afford
-    assert swap_size * 0.9 < partitions[0]['size'] <= swap_size
-    # Data of at least the requested size
-    assert partitions[1]['size'] >= data_size
+    # Hand wavey swap size test
+    assert int(partitions[0]['size'] / (1024 ** 3) + 0.5) == 2
 
+    # The data partition should start after a 'grain_size' gap
+    assert partitions[1]['start_sector'] == partitions[0]['end_sector'] + grain_size
 
-def test_disk_format_with_size_with_swap_overflow_no_swap():
-    disk = call('disk.get_unused')[0]
-    disk_size = disk['size']
-    disk = disk['name']
+    # and be a modulo of grain_size
+    assert (partitions[1]['end_sector'] - partitions[1]['start_sector']) % grain_size == 0
 
-    swap_size = 512 * 1024 * 1024  # Swaps of less than 1 GiB are not accepted
+    # and be maximal sized
+    assert partitions[1]['end_sector'] >= dd['blocks'] - grain_size
 
-    data_size = disk_size - swap_size
-    call('disk.format', disk, data_size, 2)
+    # And does not clobber the MBR data at the end
+    assert partitions[0]['end_sector'] < dd['blocks'] - MBR_SECTOR_GAP
 
-    partitions = call('disk.list_partitions', disk)
-    assert len(partitions) == 1
-    # Data of at least the requested size
-    assert partitions[0]['size'] >= data_size
+    # Hand-wavy data size test
+    assert partitions[1]['size'] > (disk['size'] - partitions[0]['size']) * 0.99
 
 
-def test_disk_format_removes_existing_partition_table():
-    disk = call('disk.get_unused')[0]['name']
+@pytest.mark.parametrize("swap_val", [-10, 2.5, 1024])
+def test_disk_format_with_invalid_swap(unused_disk, swap_val):
+    """
+    Confirm we can handle erroneous input
+    """
+    assert unused_disk[2] != 0, 'ERROR: Should not run this test without a non-zero grain_size'
+    disk = unused_disk[0]
 
-    call('disk.format', disk, None, 2)
-    call('disk.format', disk, None, 0)
+    with pytest.raises(CallError) as e:
+        call('disk.format', disk['name'], swap_val)
+
+    # The error response is input dependent
+    if swap_val > 100:
+        assert e.value.errmsg == (
+            f'Disk {disk["name"]!r} capacity is too small. '
+            'Please use a larger capacity drive or reduce swap.'
+        )
+    else:
+        assert e.value.errmsg == (
+            'Requested swap must be a non-negative integer'
+        )
+
+
+def test_disk_format_removes_existing_partition_table(unused_disk):
+    """
+    Confirm we can repartion
+    """
+    assert unused_disk[2] != 0, 'ERROR: Should not run this test without a non-zero grain_size'
+    disk = unused_disk[0]
+
+    call('disk.format', disk['name'], 2)
+    call('disk.format', disk['name'], 0)
```

Comment:
Maybe call `disk.list_partitions` after each of these and do a basic sanity check (e.g. partition count / type check).